core/src/main/java/org/apache/calcite/rel/rules/SubQueryRemoveRule.java

```diff
@@ -730,18 +731,31 @@ private static RexNode rewriteIn(RexSubQuery e, Set<CorrelationId> variablesSet,
       dtAlias = "dt" + subQueryIndex;
     }
     builder.as(dtAlias);
-    int refOffset = offset;
-    final List<RexNode> conditions =
-        Pair.zip(expressionOperands, builder.fields()).stream()
-            .map(pair -> builder.equals(pair.left, RexUtil.shift(pair.right, refOffset)))
-            .collect(Collectors.toList());
+    final int refOffset = prevInputSize + offset;
     switch (logic) {
     case TRUE:
-      builder.join(JoinRelType.INNER, builder.and(conditions), variablesSet);
-      return trueLiteral;
+      List<RexNode> prevInputConditions = new ArrayList<>(expressionOperands.size());
+      List<RexNode> otherConditions = new ArrayList<>(expressionOperands.size());
+      for (Pair<RexNode, RexNode> pair : Pair.zip(expressionOperands, builder.fields())) {
+        if (pair.left.isA(SqlKind.INPUT_REF)
+            && ((RexInputRef) pair.left).getIndex() >= (totalOffset - prevInputSize)) {
+          prevInputConditions.add(
+              builder.equals(RexUtil.shift(pair.left, prevInputSize - totalOffset),
```

Comment:
Please store pair.left & pair.right in a variable & use it at all the places

Reply:
Changed as suggested